### PR TITLE
[feat] Add Monitoring

### PR DESCRIPTION
**Pull request by warittornc**
## Implementation details
- Do 1
- Do 2

## Please ensure the following requirements are met before submitting a pull request:

- [ ] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the GitHub PR explorer)

### CODE REVIEW

File: bothan-api/server/src/config/monitoring.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MonitoringConfig {
+    #[serde(default = "default_endpoint")]
+    pub endpoint: String,
+    /// The path to where the key for the monitoring service is stored.
+    #[serde(default = "default_path")]
+    pub path: PathBuf,
+}
+
+// TODO: add default endpoint once finalized
```
Comment (colmazia) [inline bothan-api/server/src/config/monitoring.rs:14]: Don't forget na

File: bothan-api/server/src/api/crypto.rs [hunk 2]
```diff
@@ -125,7 +152,22 @@ impl PriceService for CryptoQueryServer {
...
+            match monitoring_client.post_price(cloned_prices).await {
+                Ok(r) => match r.status() {
+                    reqwest::StatusCode::OK => info!("successfully sent data to monitoring"),
+                    _ => error!("failed to send data to monitoring: {:?}", r.text().await),
+                },
+                Err(e) => error!("failed to send data to monitoring: {}", e),
+            }
+        });
 
         Ok(Response::new(GetPricesResponse { prices }))
```
Comment (colmazia) [inline bothan-api/server/src/api/crypto.rs:172]: We should send the uuid along with prices for Grogu

File: bothan-api/server/src/monitoring/client.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    pub async fn post_price(&self, prices: Vec<Price>) -> Result<Response, PostError> {
+        self.post(Topic::Price, prices).await
+    }
+
+    pub async fn post_heartbeat(
+        &self,
+        active_signal_ids: ActiveSignalIDs,
+    ) -> Result<Response, PostError> {
+        let bothan_info = BothanInfo::new(
+            VERSION.to_string(),
```
Comment (colmazia) [inline bothan-api/server/src/monitoring/client.rs:39]: Just a question, is version alone enough to derive how many sources this validator support?

Comment (warittornc) [reply bothan-api/server/src/monitoring/client.rs:39]: no, this is just the current bothan version that the validator is running

**Review by RogerKSI — CHANGES_REQUESTED** [review]

**Review by RogerKSI — APPROVED** [review]

File: bothan-api/server-cli/src/commands/start.rs [hunk 4]
```diff
@@ -139,18 +171,22 @@ async fn init_crypto_server(
...
     let stale_threshold = config.manager.crypto.stale_threshold;
     let version_req = VersionReq::from_str(&format!("<={}", VERSION))
         .with_context(|| "Failed to parse bothan version requirement")?;
     let mut manager =
         CryptoAssetInfoManager::new(manager_store, ipfs_client, stale_threshold, version_req);
-
     init_crypto_workers(&mut manager, &store, &config.manager.crypto.source).await?;
 
-    Ok(CryptoQueryServer::new(manager))
+    let monitoring_client = init_monitoring_client(config).await?;
```
Comment (taobun) [inline bothan-api/server-cli/src/commands/start.rs:184]: Move to main

File: bothan-api/server/src/monitoring/client.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    pub async fn post_heartbeat(
+        &self,
+        uuid: String,
+        active_signal_ids: ActiveSignalIDs,
+    ) -> Result<Response, PostError> {
+        let bothan_info = BothanInfo::new(
+            VERSION.to_string(),
+            active_signal_ids.into_iter().collect::<Vec<String>>(),
```
Comment (colmazia) [inline bothan-api/server/src/monitoring/client.rs:43]: So should we add supported sources in this too?

Comment (warittornc) [reply bothan-api/server/src/monitoring/client.rs:43]: I think we've decided to keep that in telemetry instead rather than the public monitoring section

**Review by colmazia — APPROVED** [review]

**Review by RogerKSI — APPROVED** [review]

**Pull request merged**